basket/news/tasks.py

```diff
@@ -717,6 +717,157 @@ def record_source_url(email, source_url, newsletter_id):
+
+    invoice.finalized *should* be sent first, which is missing the following
+    credit card related fields:
+
+    - brand
+    - last4
+    - exp_month
+    - exp_year
+    """
+
+    # determine if new or recurring payment
+    recurring_charge = True if data['current_period_start'] and data['current_period_end'] else False
+
+    if recurring_charge:
+        statsd.incr('news.tasks.process_subhub_event.new_subscription_charge')
+    else:
+        statsd.incr('news.tasks.process_subhub_event.recurring_subscription_charge')
+
+    # determine which event we're receiving
+    event = 'succeeded' if 'brand' in data and 'last4' in data else 'finalized'
+
+    # try to find the opportunity (in case one of the events already fired)
+    user_data = get_user_data(stripe_id=data['customer_id'])
+
+    if user_data:
+        # see if an opportunity exists with the provided charge_id
+        charge_id = data['charge_id']
+
+        transaction_data = {
+            'Stripe_Cust_Id__c': data['customer_id'],
+            'PMT_Subscription_ID__c': data['subscription_id'],
+            'CloseDate': iso_format_unix_timestamp(data['created']),
+            'Billing_Cycle_End__c': iso_format_unix_timestamp(data['period_end']),
+            'Billing_Cycle_Start__c': iso_format_unix_timestamp(data['period_start']),
+            'Amount': data['amount_paid'],  # TODO: *may* need to divide by 100 to get dollars - check w/marty
+            'currency__c': data['currency'],
+            'Name': 'Subscription Services',
+            'RecordTypeId': settings.SUBHUB_OPP_RECORD_TYPE,  # 012R0000000dQ4dIAE for dev
+            'StageName': 'Closed Won',
+            'Payment_Source__c': 'Stripe',
+            'Invoice_Number__c': data['invoice_id'],
+            'Processors_Fee__c': data['application_fee_amount'],
+            'Net_Amount__c': ''  # TODO: check w/marty if we are getting this or if we need to do math (amount_paid - application_fee_amount)
+            'Conversion_Amount__c': ''  # TODO: check w/marty if we need this
+        }
+
+        if event == 'succeeded':
+            transaction_data['Credit_Card_Type__c'] = data['brand']
+            transaction_data['Last_4_Digits__c'] = data['last4']
+            transaction_data['Credit_Card_Exp__c'] = data['exp_month']
+            transaction_data['Credit_Card_Exp__c'] = data['exp_year']  # TODO: check w/marty on SF field names here - they are duplicated in the doc
+
+        try:
+            # will raise a SalesforceMalformedRequest if not found
+            sfdc.opportunity.update('PMT_Transaction_ID__c/{}'.format(charge_id), transaction_data)
+            # TODO: is this the proper syntax? unsure about the use/purpose of the first param here - is there some magic happening that makes this a key?
+        except sfapi.SalesforceMalformedRequest as e:
+            sfdc.opportunity.create(transaction_data)
+    else:
+        # TODO: do we need to do anything here if the user wasn't found?
```

Comment:
Maybe there's enough info in the data to create the user if none is found? Really only need email address.

Reply:
According to the doc, the only user identifiable info in these payloads is the Stripe ID, so I don't think we can do much?